### ISSUE TITLE
Forward Agent does not work

### ISSUE THREAD

**Issue by beliys** [asbru-cm/asbru-cm#519]
**Describe the bug**
Forward Agent does not work.

**To Reproduce**
1. Go to 'Edit Connection' for ssh connection  
2. Enable "Forward Agent" on 'SSH Connection'
3. Run local `eval $(ssh-agent)` 
4. Run local `ssh-add` 
5. Connect to server by ssh
6. Run on remote server `ssh-add -L` and look error "Could not open a connection to your authentication agent."

**Desktop (please complete the following information):**
    OS: Manjaro KDE
    Version: latest

**Comment by gfrenoy** [comment]
I would think this is because it is using variable environment for this.

Can you try to run step 2 and 3 *before* starting Ásbrú (and start Ásbrú from the same shell) ?

**Comment by beliys** [comment]
@gfrenoy Thank you, this works, but this is not a very convenient way.

**Comment by gfrenoy** [comment]
I understand this perfectly but I do fear this is the way it works and I don't see how we can offer a better user experience.

**Comment by hanspr** [comment]
@gfrenoy 

Pardon my ignorance on this, but what would happen if we check for the `-A` flag, and check the existence of the environment variable that you mention, and based on those; can we launch those start up commands inside asbru_conn? or request PACTerminal to start them up one time?

Which command(s) have to be executed (both, one of them)

```
eval $(ssh-agent)
ssh-add
```

What is the environment variable to look at?

**Comment by gfrenoy** [comment]
Yes, that could help but ```ssh-add``` may require some additional parameters (like the private key you want to use) and ask your password interactively so it might not be trivial but definitely a better experience :)

On some systems (at least mine ;p), you also have an agent that is started automatically for accessing the OS "keyring".

Just call ```ssh-agent``` to discover the variable that are created, it should be ```SSH_AUTH_SOCK``` and ```SSH_AGENT_PID```.

And, another approach would maybe to use some pre-exec commands, isn't it ?

**Comment by beliys** [comment]
some documentation from arch wiki - https://wiki.archlinux.org/index.php/SSH_keys#SSH_agents

**Comment by hanspr** [comment]
> And, another approach would maybe to use some pre-exec commands, isn't it ?

Suppose, that could be a good option, I guess that is the intention of that thing

**Comment by hanspr** [comment]
@beliys 

Thanks for the link, this is interesting.

![imagen](https://user-images.githubusercontent.com/1572396/78176264-738f5880-7419-11ea-8e85-6c8922d51571.png)

In my machine.

![imagen](https://user-images.githubusercontent.com/1572396/78176773-4becc000-741a-11ea-808e-775ce271146a.png)

https://keepassxc.org/docs/#faq-ssh-agent-how

![imagen](https://user-images.githubusercontent.com/1572396/78176580-f1536400-7419-11ea-987a-b82feb0f6ba3.png)

If I understand this correctly if I have keepassxc installed and running from startup, I already have an ssh-agent running.

**Comment by gfrenoy** [comment]
> If I understand this correctly if I have keepassxc installed and running from startup, I already have an ssh-agent running.

One more condition is to check the box :-)

![image](https://user-images.githubusercontent.com/981546/78219238-6fb61180-74bf-11ea-8d1b-dfc077e8e7fa.png)

But trick remains the same ; how to find out the environment variable of another session ; and if you have several agents, which one to pick ?

At this stage, we can probably use this thread to document how one can set that up but not more ; don't you think ?

**Comment by hanspr** [comment]
Yes, I need to familiarize with the concept, I've never used it.

**Comment by gfrenoy** [comment]
Is there anything else we can do here ?

**Comment by beliys** [comment]
we can use the Forward Agent with asbru-cm, but we must manually prepare a separate terminal and run asbru-cm in it. This is not very convenient...

**Comment by hanspr** [comment]
> some documentation from arch wiki - https://wiki.archlinux.org/index.php/SSH_keys#SSH_agents

If I understood this documentation correctly, the forward agent can be started on the login session and then would be available for asbru too, no?

![imagen](https://user-images.githubusercontent.com/1572396/81778333-42925100-94b8-11ea-9c7e-6732d9a73b60.png)

**Comment by beliys** [comment]
I checked, but it did not work for me((

**Comment by mooky31** [comment]
Hello,

I ran into that problem today, and found the solution thanks to this thread.
Thank you guys.

I automated the agent startup addig this to my `~/.bashrc`:
```
# SSH agent
eval $(ssh-agent)
ssh-add ~/.ssh/privkey
```

Adding an option to Asbru to automatically start the agent and load the keys would be nice though.

BTW, this SSH client is really great. Kudos to the devs !

**Comment by beliys** [comment]
@mooky31 Your method has several disadvantages:
1) for each new terminal, a separate process will be created that will not close together with the terminal instead of using only 1 process.
2) your method does not guarantee that it will work in asbru-cm on all systems, since .bashrc is not always executed for login shells.

To solve 1 problem, you need to either write scripts to reuse the existing ssh-agent or use utilities that can do it out of the box.
To solve problem 2, you need to either ensure that `.bashrc` is executed for login shells or additionally add to `.bash_profile`.
For me, it turned out to be optimal to use the `keychain` utility, which solves 1 problem very well and works very stable.
I just added command `eval $(keychain --eval --agents ssh --quiet id_rsa)` to `.bash_profile` and `.bashrc.`
This completely solved all the inconveniences associated with using ssh-agent, but the integration of asbru-cm with password managers that can work as ssh-agent (for example gnome keyring/KeePassXC/KWallet) would definitely not be superfluous.

P.S. Sorry for not writing about this method earlier.

**Comment by mooky31** [comment]
> 
> 
> @mooky31 Your method has several disadvantages:
> 
>     1. for each new terminal, a separate process will be created that will not close together with the terminal instead of using only 1 process.
> 
>     2. your method does not guarantee that it will work in asbru-cm on all systems, since .bashrc is not always executed for login shells.
> 
> 
> To solve 1 problem, you need to either write scripts to reuse the existing ssh-agent or use utilities that can do it out of the box.
> To solve problem 2, you need to either ensure that `.bashrc` is executed for login shells or additionally add to `.bash_profile`.
> For me, it turned out to be optimal to use the `keychain` utility, which solves 1 problem very well and works very stable.
> I just added command `eval $(keychain --eval --agents ssh --quiet id_rsa)` to `.bash_profile` and `.bashrc`.
> This completely solved all the inconveniences associated with using ssh-agent, but the integration of asbru-cm with password managers that can work as ssh-agent (for example gnome keyring/KeePassXC/KWallet) would definitely not be superfluous.
> 
> P.S. Sorry for not writing about this method earlier.

You're right, thanks for the better solution !

**Comment by gfrenoy** [comment]
So is there anything to do in Ásbrú about this ?  Shall we close the issue ?  Thanks.

**Comment by beliys** [comment]
@gfrenoy As I said earlier, the integration of asbru-cm with password managers that can work as ssh-agent (for example gnome keyring/KeePassXC/KWallet) would definitely not be superfluous.

**Comment by chodorenko** [comment]
I already have ssh-agent which was launched at OS start

chodorenko@chodorenko: ~> set | grep SSH_A
SSH_AGENT_PID = 3395
SSH_ASKPASS = /usr/lib/ssh/ssh-askpass
SSH_AUTH_SOCK = /tmp/ssh-RgEYM7vgNqdD/agent.3337

A key was loaded into it(key protected by passwd), for what reason does asbru-cm not read these variables? Why would I run 100,500 additional agents?

**Comment by gfrenoy** [comment]
@chodorenko It should work but you need to be sure the environment variables are correctly set when you start Ásbú.  According to the way you are starting Ásbrú and the SSH agent, it might not be obvious...

See https://github.com/asbru-cm/asbru-cm/issues/519#issuecomment-627768655 ; not convenient but it works.
